airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py

```diff
@@ -328,101 +339,133 @@ def create_labels_for_pod(context) -> dict:
+        result = self.launcher.extract_xcom(pod)
+        self.log.info(result)
+        return json.loads(result)
 
-            launcher = self.create_pod_launcher()
+    def execute(self, context):
+        remote_pod = None
+        try:
+            self.pod_request_obj = self.build_pod_request_obj(context)
+            self.pod = self.get_or_create_pod(  # must set `self.pod` for `on_kill`
+                pod_request_obj=self.pod_request_obj,
+                context=context,
+            )
+            self.await_pod_start(pod=self.pod)
 
-            if len(pod_list.items) == 1:
-                try_numbers_match = self._try_numbers_match(context, pod_list.items[0])
-                final_state, remote_pod, result = self.handle_pod_overlap(
-                    labels, try_numbers_match, launcher, pod_list.items[0]
+            if self.get_logs:
+                self.launcher.follow_container_logs(
+                    pod=self.pod,
+                    container_name=self.BASE_CONTAINER_NAME,
                 )
             else:
-                self.log.info("creating pod with labels %s and launcher %s", labels, launcher)
-                final_state, remote_pod, result = self.create_new_pod_for_operator(labels, launcher)
-            if final_state != State.SUCCESS:
-                raise AirflowException(f'Pod {self.pod.metadata.name} returned a failure: {remote_pod}')
-            context['task_instance'].xcom_push(key='pod_name', value=self.pod.metadata.name)
-            context['task_instance'].xcom_push(key='pod_namespace', value=self.namespace)
-            return result
-        except AirflowException as ex:
-            raise AirflowException(f'Pod Launching failed: {ex}')
+                self.launcher.await_container_completion(
+                    pod=self.pod, container_name=self.BASE_CONTAINER_NAME
+                )
 
-    def handle_pod_overlap(
-        self, labels: dict, try_numbers_match: bool, launcher: Any, pod: k8s.V1Pod
-    ) -> Tuple[State, k8s.V1Pod, Optional[str]]:
-        """
+            if self.do_xcom_push:
+                result = self.extract_xcom(pod=self.pod)
+            remote_pod = self.launcher.await_pod_completion(self.pod)
+        finally:
+            self.cleanup(
+                pod=self.pod or self.pod_request_obj,
+                remote_pod=remote_pod,
+            )
+        if self.do_xcom_push:
+            ti = context['ti']
+            if remote_pod:
+                ti.xcom_push(key='pod_name', value=remote_pod.metadata.name)
+                ti.xcom_push(key='pod_namespace', value=remote_pod.metadata.namespace)
+            return result
 
-        In cases where the Scheduler restarts while a KubernetesPodOperator task is running,
-        this function will either continue to monitor the existing pod or launch a new pod
-        based on the `reattach_on_restart` parameter.
+    def cleanup(self, pod, remote_pod):
```

Comment:
Are these both V1Pod types?

Reply:
i believe so.  i think one is the object that the request is made with and the other is after we've read status.  are you thinking "please add type hints"?  or thinking "do we really need both"?

Reply:
Please add type hints :)